src/relay/backend/contrib/csource/codegen.cc

```diff
@@ -0,0 +1,229 @@
+  /*! \brief The declaration statements of a C compiler compatible external function. */
+  std::vector<std::string> func_decl_;
+  /*! \brief The declaration statements of buffers. */
+  std::vector<std::string> buf_decl_;
+  /*! \brief The name and index pairs for output. */
+  std::vector<std::pair<std::string, int>> out_;
+};
+
+class CSourceCodegen : public ExternCodegenBase {
+ public:
+  void CreateExternFunction(const Function& func) {
+    CHECK(func.defined())
+        << "Input error: external codegen expects a Relay function.";
+
+    // Record subgraph ID for runtime invoke.
+    auto sid = GetSubgraphID(func, "ccompiler");
+
+    auto builder = CSourceBuilder("ccompiler_" + sid);
+    builder.VisitExpr(func->body);
+    code_stream_ << builder.JIT();
+  }
+
+  runtime::Module CreateExternModule(const NodeRef& ref) {
+    // Create headers
+    code_stream_ << "#include <cstdint>\n";
+    code_stream_ << "#include <iostream>\n";
+    code_stream_ << "#include <cstdlib>\n";
+    code_stream_ << "#include <stdio.h>\n";
+    code_stream_ << "#include <cstring>\n";
+    code_stream_ << "#include <tvm/runtime/c_runtime_api.h>\n";
+    code_stream_ << "#include <dlpack/dlpack.h>\n";
+
+    // Append some common macro for operator definition.
+    const char* operator_macro = R"op_macro(
+    #define CSOURCE_BINARY_OP_1D(p_ID_, p_OP_, p_DIM1_)       \
+      extern "C" void p_ID_(float* a, float* b, float* out) { \
+        for (int64_t i = 0; i < p_DIM1_; ++i) {               \
+          out[i] = a[i] p_OP_ b[i];                           \
+        }                                                     \
+      }
+    
+    #define CSOURCE_BINARY_OP_2D(p_ID_, p_OP_, p_DIM1_, p_DIM2_)  \
+      extern "C" void p_ID_(float* a, float* b, float* out) {     \
+        for (int64_t i = 0; i < p_DIM1_; ++i) {                   \
+          for (int64_t j = 0; j < p_DIM2_; ++j) {                 \
+            int64_t k = i * p_DIM2_ + j;                          \
+            out[k] = a[k] p_OP_ b[k];                             \
+          }                                                       \
+        }                                                         \
+      }
+    )op_macro";
+
+    code_stream_ << operator_macro << "\n\n";
+
+    if (ref->IsInstance<FunctionNode>()) {
+      CreateExternFunction(Downcast<Function>(ref));
+    } else if (ref->IsInstance<relay::ModuleNode>()) {
+      relay::Module mod = Downcast<relay::Module>(ref);
+      for (const auto& it : mod->functions) {
+        CreateExternFunction(Downcast<Function>(it.second));
+      }
+    } else {
+      LOG(FATAL) << "The input ref is expected to be a Relay function or module"
+                 << "\n";
+    }
+
+    // Create a CSourceModule
+    const auto* pf = runtime::Registry::Get("module.csource_module_create");
+    CHECK(pf != nullptr) << "Cannot find csource module to create the external function";
+    return (*pf)(code_stream_.str(), "cc");
```

Comment:
@zhiics @comaniac can you tell me where the actual compilation of c code happens? Do we get DSOModule as a result?

Reply:
The compilation happens when exporting the library. You could refer to our unit test example: https://github.com/apache/incubator-tvm/pull/4482/files#diff-eeeb210b0f9899954aa836dd05e94c39R32-R40

As you can see in this example, by importing the external function to the normal DSO module, users will now only have a single DSO module.

Reply:
I see, thanks. Makes sense.